packages/SwingSet/src/kernel/vatManager/supervisor-subprocess-xsnap.js

```diff
@@ -0,0 +1,237 @@
+  encode: tagged => encoder.encode(JSON.stringify(tagged)).buffer,
+
+  /** @type { (msg: ArrayBuffer) => Tagged } */
+  decode(msg) {
+    const txt = decoder.decode(msg);
+
+    /** @type { Tagged } */
+    const item = JSON.parse(txt);
+    assert(Array.isArray(item), details`expected array`);
+    assert(item.length > 0, details`empty array lacks tag`);
+    return item;
+  },
+};
+
+function makeConsole(_tag) {
+  const log = console; // TODO: anylogger(tag);
+  const cons = {
+    debug: log.debug,
+    log: log.log,
+    info: log.info,
+    warn: log.warn,
+    error: log.error,
+  };
+  return harden(cons);
+}
+
+function testLog(...args) {
+  // @ts-ignore xsnap provides issueCommand global
+  // eslint-disable-next-line no-undef
+  issueCommand(Item.encode(['testLog', ...args]));
+}
+
+/**
+ * @param { (value: void) => void } f
+ * @param { string } errmsg
+ */
+function runAndWait(f, errmsg) {
+  Promise.resolve()
+    .then(f)
+    .then(undefined, err => workerLog(`doProcess: ${errmsg}:`, err));
+  return waitUntilQuiescent();
+}
+
+function makeWorker() {
+  /** @type { Record<string, (...args: unknown[]) => void> | null } */
+  let dispatch = null;
+
+  /** @type { (dr: Tagged, errmsg: string) => Promise<Tagged> } */
+  async function doProcess(dispatchRecord, errmsg) {
+    assert(dispatch);
+    const theDispatch = dispatch;
+    const [dispatchOp, ...dispatchArgs] = dispatchRecord;
+    assert(typeof dispatchOp === 'string');
+    workerLog(`runAndWait`);
+    await runAndWait(() => theDispatch[dispatchOp](...dispatchArgs), errmsg);
+    workerLog(`doProcess done`);
+    /** @type { Tagged } */
+    const vatDeliveryResults = harden(['ok']);
+    return vatDeliveryResults;
+  }
+
+  /** @type { (ts: unknown, msg: any) => Promise<Tagged> } */
+  function doMessage(targetSlot, msg) {
+    const errmsg = `vat[${targetSlot}].${msg.method} dispatch failed`;
+    return doProcess(
+      ['deliver', targetSlot, msg.method, msg.args, msg.result],
+      errmsg,
+    );
+  }
+
+  /** @type { (pv: unknown, msg: any) => Promise<Tagged> } */
+  function doNotify(primaryVpid, resolutions) {
+    const errmsg = `vat.promise[${primaryVpid}] failed`;
+    return doProcess(['notify', primaryVpid, resolutions], errmsg);
+  }
+
+  /**
+   * @param {unknown} vatID
+   * @param {unknown} bundle
+   * @param {unknown} vatParameters
+   * @param {unknown} virtualObjectCacheSize
+   * @returns { Promise<Tagged> }
+   */
+  async function setBundle(
+    vatID,
+    bundle,
+    vatParameters,
+    virtualObjectCacheSize,
+  ) {
+    /** @type { (item: Tagged) => unknown } */
+    function doSyscall(vatSyscallObject) {
+      return JSON.parse(
+        decoder.decode(
+          // @ts-ignore
+          // eslint-disable-next-line no-undef
+          issueCommand(Item.encode(['syscall', ...vatSyscallObject])),
+        ),
+      );
+    }
+
+    const syscall = harden({
+      send: (...args) => doSyscall(['send', ...args]),
```

Comment:
Like the `handleCommand` and `issueCommand` duals, I think `sysCall` and `answerSysCall` are intuitive duals. `doSyscall`’s got two verbs. There’s also a similar distinction here between layers that might be better captured by `sysCall` vs `taggedSysCall`.

Reply:
`doSyscall` is patterned after other vat worker code. I'd rather not change it here without changing it in the other workers. I'll leave it as is unless / until I hear otherwise from @warner .

Reply:
yeah, we currently expose a `syscall` object (with a variety of methods) to the liveslots layer of the vat. Then each time that is used, we collect the type and arguments into a "vat syscall object" (a serializable chunk of data) which can be sent over the worker-kernel interface. One could argue that the liveslots layer should create the "vat syscall object" data directly, rather than being given a handful of methods to invoke.. I don't think I have strong opinions either way, but I can imagine that the type annotations would be more meaningful in one mode than the other.

@dckc a head's up here, @FUDCo 's work in #2242 and beyond will change the methods on this object: the three `fulfillToData`/`fulfillToPresence`/`reject` methods are merging into a single method.

Reply:
only `doNotify` has changed so far.

Reply:
spoke too soon. Chip landed the other bit.